### ISSUE TITLE
"PRESAGE NOT AVAILABLE--Switching to NGram mode" with presage_wcf_service_system_tray.exe running

### ISSUE THREAD

**Issue by gabrielbenjamin** [OptiKey/OptiKey#349]
I started presage_wcf_service_system_tray.exe twice, once before starting OptiKey and once after, and upon restart received the error above. Have I made a mistake?

**Comment by zelmon64** [comment]
Which version of Presage did you install? It will only work with `presage-0.9.2~beta20150909-32bit`. Did you let it install to it's default location in `Program Files (x86)` or did you specify a different location or manually move it after installation?

**Comment by gabrielbenjamin** [comment]
The very same version, installed with all the defaults. And it's the latest version of OptiKey. I reinstalled Presage, tried to use it as-is with OptiKey, even used text2ngram beforehand, but I always get that error message.

**Comment by zelmon64** [comment]
Could you please enable Presage then post your resulting `\AppData\Roaming\JuliusSweetland\OptiKey\Logs\OptiKey.log` file?

**Comment by JuliusSweetland** [comment]
@gabrielbenjamin An issue has been identified with a user installing the 64 bit version of Presage. Please can you check whether you have the 32 or 64 bit version of Presage 0.9.2 installed? I suspect you have the 64 bit version installed (OptiKey requires the 32 bit version). You can tell from the install location; if you are running a 64 bit operating system (which your logs say you are "OS bitness: 64-Bit") then check if it is installed in "Program Files" (you installed the 64 bit version) of "Program Files (x86)" (you have the 32 bit version).

If you have the 64 bit version please uninstall it, install the 32 bit version, and try again.

**Comment by JuliusSweetland** [comment]
@gabrielbenjamin After checking your Presage installation please try the new version of OptiKey; 2.9.0 available here https://github.com/OptiKey/OptiKey/releases/tag/v2.9.0.3

**Comment by JuliusSweetland** [comment]
Closing as I believe this should be fixed by 2.9.0 - please re-open or create a new issue if not.

**Comment by gabrielbenjamin** [comment]
@JuliusSweetland I installed Presage from the file at the link in the OptiKey docs, so I assume I installed the correct version. Installing the version of OptiKey available at the homepage, which is 2.9.0.0, for clarity. I can't download the installer at the link you posted--it says I don't have the right permissions. Thus I can't say if 2.9.0.3 fixes this.

@zelmon64 Will post the log shortly.

**Comment by JuliusSweetland** [comment]
@gabrielbenjamin I created a few versions of 2.9.0 so I may have pointed you at an old one. Let me know how the official 2.9.0 release works for you.

**Comment by gabrielbenjamin** [comment]
@zelmon64 Here it is.
[OptiKey.log.txt](https://github.com/OptiKey/OptiKey/files/1240324/OptiKey.log.txt)

@JuliusSweetland The error says otherwise, but presage-0.9.2~beta20150909-32bit is the only thing I ever installed. Something suspicious, though: I uninstalled presage and tried to delete all the related files, with the intention of reinstalling it from scratch. However, I'm told I don't have the proper permissions to delete qprompter.exe----I'm the only user on my computer, and I'm an admin. It's mystifying, but probably not an impediment. We'll see...

**Comment by gabrielbenjamin** [comment]
@JuliusSweetland No dice--same behavior. The log reads pretty much the same.

**Comment by zelmon64** [comment]
@gabrielbenjamin I see the problem. Line 2 of that log file should read 
```
... [ 1] INFO  JuliusSweetland.OptiKey.App: Presage path: C:\Program Files\presage | Presage start menu folder: presage-0.9.2~beta20150909 | OS bitness: 64-Bit
```
but instead your Presage start menu folder is `Presage`. If you reinstall it stating the start menu folder `presage-0.9.2~beta20150909` it should work.

@JuliusSweetland From this I'm wondering if the Presage version and bitness checks are too restricting. Should we instead rely more on whether `presageTestInstance` succeeds?

**Comment by JuliusSweetland** [comment]
We could relax that start folder check. Do you know what happens if you try
to use a different version (with the correct bitness) of Presage?

**Comment by zelmon64** [comment]
@JuliusSweetland I suppressed the bitness and version checks then tried presage-0.9.1-64bit (I already had it downloaded, I'll grab the 32bit version next). The new message box showed explaining there was a problem but after I closed that OptiKey seemed to freeze. I got one of those not responding messages come up. After I manually closed OptiKey and restarted it, it was hapily using NGram.

Here's the log from the unresponsive run
[OptiKey.log.txt](https://github.com/OptiKey/OptiKey/files/1241201/OptiKey.log.txt)

**Comment by zelmon64** [comment]
@JuliusSweetland Well this is interesting. OptiKey can now use `presage-0.9.1-32bit` seemingly without any problems. I had thought Presage 0.9.1 wouldn't work but perhaps I had only tested the 64bit version.

**Comment by JuliusSweetland** [comment]
I would have expected OptiKey to detect that the 64 bit version is
installed to the wrong directory ("Program Files" instead of "Program Files
(x86)"), unless you are running a 32 bit operating system?

**Comment by zelmon64** [comment]
I suppressed the bitness check as well as the version check to see wha would happen. The bitness check will fail if the user installs Presage to a non-default location.

**Comment by JuliusSweetland** [comment]
I think we should leave the checks as they are. It is up to the user to
ensure they install using default settings. What do you think?

**Comment by zelmon64** [comment]
I agree that's probably the safest thing for OptiKey. Where should it be put that the user must use the default installation directory and Start Menu folder for Presage? Should we also allow `presage-0.9.1-32bit` or force everyone to use `presage-0.9.2~beta20150909-32bit` to make sure that there are no discrepancies?

**Comment by JuliusSweetland** [comment]
I can add a note to the wiki page explaining that a default install is
required. Could you extend the start path check to support 0.9.1 if that
works also?

**Comment by zelmon64** [comment]
Yes I'm happy to do that.

I just noticed 2.9.0.3 keeps reporting there is an update available. Not a huge problem but thought you might want to know.

**Comment by JuliusSweetland** [comment]
Is that the official release or an earlier pre release?

**Comment by zelmon64** [comment]
The official release. I just re-downloaded it and it gives me the update popup at the start.

**Comment by JuliusSweetland** [comment]
Could you please try again now? I was incorrectly tagging my releases as I
was tired!

**Comment by zelmon64** [comment]
Yep that did it. No update popup message now. :)

**Comment by JuliusSweetland** [comment]
Great thanks. I'll update the wiki page shortly and I'll add that OptiKey
supports 0.9.1 when you have had a chance to add that check.

**Comment by zelmon64** [comment]
I have the presage-0.9.1 support ready but I noticed the logs still reports `Assembly version: 2.9.0.0`. Should I update that as part of my PR or leave that for you?

**Comment by JuliusSweetland** [comment]
I would have thought it always did report the version number to 4 digits,
even though I only use 3?

**Comment by zelmon64** [comment]
What I meant was that I thought it should be 2.9.0.**3** not 2.9.0.**0**.

**Comment by JuliusSweetland** [comment]
Oh right. No I set all of the assembly versions to 2.9.0 including the
press releases. I then use tags in GitHub to mark releases with a version
number which is used to check for updates.

**Comment by gabrielbenjamin** [comment]
@JuliusSweetland  Can I just rename the Start Menu Folder for Presage to the default?

**Comment by JuliusSweetland** [comment]
You could certainly try. No guarantees! If it doesn't work can you
uninstall and reinstall to the default start menu location, or does it
always default to "Presage" now?

**Comment by zelmon64** [comment]
@gabrielbenjamin the important thing is the registry key that had the name of the start menu folder. Editing that should do it but it's more risky than reinstalling Presage.

**Comment by gabrielbenjamin** [comment]
I uninstalled presage and reinstalled it with the default Start menu folder, and it works now. Permit me to be clueless, but couldn't you retrieve the name of the Start menu folder presage is in from the registry and, I don't know, put it in an environment variable? I assume the name of the registry for presage is consistent. It just seems so fragile this way.

**Comment by gabrielbenjamin** [comment]
I'm content to close this--or rather keep it closed, since I never figured out how to reopen it.

**Comment by gabrielbenjamin** [comment]
Spoke too soon--I changed presage.xml according to the wiki, and OptiKey gave me an error saying I have the wrong version of presage, then the same "PRESAGE NOT AVAILABLE" message. I couldn't revert it, but reinstalling fixed it, and the only two changes had been to presage.xml and, of course, to presage's language model. I see OptiKey2.10 is out--I'll try again with that.

**Comment by gabrielbenjamin** [comment]
Either installing OptiKey 2.10 or restarting fixed the problem. It's weird, but I guess you can leave this bug closed.

**Comment by JuliusSweetland** [comment]
Odd. I'll look into compatibility issues this evening, but I'm glad it's
finally working for you!
